### PR TITLE
hyperparam adjustments and multi runs

### PR DESCRIPTION
**Pull request by mparrett**
# Summary

Leaving this in Draft because I implemented separate run directories only to remember that trainer does that already. The rest of the code is pretty solid.

Key things to look at:

- Trained on a fresh climbs.csv created with your latest changes.
- Added a frames shuffler to create augmented datasets. Doing this as a separate utility was a less complex path forward. `python3 shuffler.py`. Remember to re-train the tokenizer.
- Hyperparameters. The hyperparameters as in this PR are working well for me (in terms of reasonable train and eval losses)
- Early stopping. We want to stop training sometime around when the eval loss starts climbing (even if train loss is still dropping). On my laptop, this currently happens around 9-10 minutes into training.
- More rigorous attempt to make the alternate generator behave like pipeline and validate that it's not doing anything super weird. Didn't play around much with top_k/top_p sampling, since there are other things to stabilize first.
- Penalty debugging. Found a couple of things w.r.t. penalties that are worth a look.

```
# Note: Repeated padding tokens are bad for penalty
# also:
#     'angle_tokens': [
#        ('a65', 1056),
#        ('<pad>', 3),
```

image

Most recent and stable run is the orange line. Including a previous run for comparison, where the model quickly overfit.

Early stopping might be a little aggressive now because I am trying to iterate quickly, but also seems like grad_norm is blowing up. In any case, should eventually make sure we are looking at a smoothed/aggregate loss, since train/eval values can bounce all over the place when not looking at the aggregate. But right now when eval starts trending up, I feel pretty confident we're overfitting our small dataset.

E.g. 
image

### CODE REVIEW

**Comment by mparrett** [comment]
cc: @rparrett

**Comment by rparrett** [comment]
Thanks! Excited to take a look. Hope you're getting some form of enjoyment out of this :)

I also had a small play with shuffling a while back https://github.com/rparrett/kilter_brain_gen/commits/shuffle-frames/ where I also manually added duplicated data to the csv file (but shuffled when loading the data). Results were bizarre enough that I figured I screwed something up. edit: oh, I put a link in the readme already

**Comment by rparrett** [comment]
I think that the main concern on my mind when I was playing around with shuffling was prompts like:

`a20d20p1143r12p1162r12p1394r14`

Where a start and finish are already present and maybe a middle hold are present, and I want to fill in the rest of the climb. Sometimes this works okay, but it's mostly unclimbable nonsense. It'll toss in a few footholds and leave a 10ft gap to the finish or whatever.

**Comment by rparrett** [comment]
> Leaving this in Draft because I implemented separate run directories only to remember that trainer does that already.

Ah yeah, I think run name should be as easy as setting `TrainingArguments.run_name`.

**Comment by rparrett** [comment]
> ```
> # also:
> #     'angle_tokens': [
> #        ('a65', 1056),
> #        ('<pad>', 3),
> ```

oof, good catch.

**Comment by mparrett** [comment]
Had this bit of insight last night about constrained generation after reading your comment above. Check out b7fb29118d15cc1d69082a81ff83f79e04b4a743. It's kinda awkwardly implemented, but promising-ish results? Not sure why repeated_tokens is still happening.

```
just run src/climb_clm/generate.py
latest_dir: models/climb_clm_run_003
Checkpoint dir: models/climb_clm_run_003/checkpoint-4400
Tokenizer: models/climb_clm_tokenizer
Stats:

('a40d15', 'multinomial_with_prefix_allow', 15, 0.1)
models/climb_clm_run_003/checkpoint-4400.1.0.a40.d15,p1127r12p1387r14p1199r13p1252r13p1269r13p1521r15p1521r15p1212r13p1247r13p1299r13
repeated_tokens

('a10d10', 'multinomial_with_prefix_allow', 15, 0)
models/climb_clm_run_003/checkpoint-4400.2.0.a10.d10,p1153r12p1392r14p1469r15p1283r13p1234r13p1304r13p1454r15p1496r15p1285r13p1218r13


('a10d10', 'multinomial_with_prefix_allow', 15, 0)
models/climb_clm_run_003/checkpoint-4400.2.2.a10.d10,p1151r12p1383r14p1270r13p1337r13p1543r15p1232r13p1267r13p1281r13p1508r15p1539r15


('a20d20', 'multinomial_with_prefix_allow', 16, 0)
models/climb_clm_run_003/checkpoint-4400.0.1.a20.d20,p1147r12p1386r14p1247r13p1516r15p1347r13p1200r13p1308r13p1451r15p1456r15p1336r13p1334r13


('a40d15', 'multinomial_with_prefix_allow', 16, 0.1)
models/climb_clm_run_003/checkpoint-4400.1.1.a40.d15,p1127r12p1389r14p1457r15p1336r13p1319r13p1521r15p1353r13p1332r13p1521r15p1219r13p1233r13
repeated_tokens

('a40d15', 'multinomial_with_prefix_allow', 16, 0.2)
models/climb_clm_run_003/checkpoint-4400.1.2.a40.d15,p1191r12p1384r14p1203r13p1299r13p1250r13p1250r13p1269r13p1283r13p1269r13p1320r13p1454r15
repeated_tokens

('<s>', 'multinomial_with_prefix_allow', 16, 0.2)
models/climb_clm_run_003/checkpoint-4400.3.0.a20.d25,p1154r12p1390r14p1334r13p1333r13p1371r13p1371r13p1266r13p1334r13p1335r13p1081r15
repeated_tokens

('<s>', 'multinomial_with_prefix_allow', 16, 0.2)
models/climb_clm_run_003/checkpoint-4400.3.1.a35.d22,p1185r12p1391r14p1231r13p1271r13p1335r13p1340r13p1321r13p1340r13p1340r13p1516r15
repeated_tokens

('<s>', 'multinomial_with_prefix_allow', 16, 0.2)
models/climb_clm_run_003/checkpoint-4400.3.2.a45.d21,p1134r12p1379r14p1216r13p1266r13p1470r15p1454r15p1216r13p1470r15p1508r15p1531r15
repeated_tokens

('a20d20', 'multinomial_with_prefix_allow', 17, 0)
models/climb_clm_run_003/checkpoint-4400.0.0.a20.d20,p1200r12p1389r14p1266r13p1332r13p1452r15p1455r15p1352r13p1335r13p1247r13p1454r15p1447r15p1198r13


('a20d20', 'multinomial_with_prefix_allow', 19, 0.30000000000000004)
models/climb_clm_run_003/checkpoint-4400.0.2.a20.d20,p1143r12p1391r14p1216r13p1505r15p1332r13p1452r15p1521r15p1522r15p1456r15p1452r15p1455r15p1459r15p1459r15p15
05r15
repeated_tokens

('a10d10', 'multinomial_with_prefix_allow', 20, 0.1)
models/climb_clm_run_003/checkpoint-4400.2.1.a10.d10,p1165r12p1383r14p1283r13p1264r13p1285r13p1315r13p1254r13p1353r13p1372r13p1506r15p1452r15p1470r15p1498r15p12
85r13p1334r13
repeated_tokens
```

image

**Comment by mparrett** [comment]
> a20d20p1143r12p1162r12p1394r14

Not sure why we're exactly getting 27 tokens here, but worth a look.

('a20d20p1143r12p1162r12p1394r14', 'multinomial_with_prefix_allow', 27, 0)
models/climb_clm_run_003/checkpoint-4400.4.0.a20.d20,p1143r12p1162r12p1394r14p1254r13p1285r13p1454r15p1516r15p1540r15p1458r15p1459r15p1230r13p1264r13p1332r13p1335r13p1339r13p1231r13p1521r15p1518r15p1198r13p1456r15p1316r13p1347r13p1336r13


('a20d20p1143r12p1162r12p1394r14', 'multinomial_with_prefix_allow', 27, 0)
models/climb_clm_run_003/checkpoint-4400.4.1.a20.d20,p1143r12p1162r12p1394r14p1231r13p1237r13p1267r13p1339r13p1447r15p1531r15p1494r15p1459r15p1470r15p1512r15p1522r15p1505r15p1230r13p1270r13p1357r13p1330r13p1332r13p1467r15p1493r15p1304r13


('a20d20p1143r12p1162r12p1394r14', 'multinomial_with_prefix_allow', 27, 0)
models/climb_clm_run_003/checkpoint-4400.4.2.a20.d20,p1143r12p1162r12p1394r14p1352r13p1228r13p1336r13p1283r13p1334r13p1451r15p1452r15p1456r15p1505r15p1454r15p1335r13p1459r15p1462r15p1332r13p1447r15p1449r15p1299r13p1338r13p1267r13p1232r13

**Comment by rparrett** [comment]
Those are a little funny but a slight improvement, I think.

File: src/climb_clm/generate.py [hunk 1]
```diff
@@ -1,26 +1,40 @@
 import re
 import sys
+from pprint import pprint
+from rich.console import Console
 
+from util import print_colored_role_names
```
Comment (rparrett) [inline src/climb_clm/generate.py:6]: Looks like this module isn't present here

**Comment by mparrett** [comment]
Thanks. Added and pushed to branch.
